File: src/knight-tour/src/KnightTour.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#include <sys/time.h>

#define DEFAULT_BOARD_SIDE_SIZE     (8)
#define DEFAULT_START_ROW           (0)
#define DEFAULT_START_COLUMN        (0)
#define DEFAULT_ITERATIONS          (1)

struct Move {
    int rowDelta;
    int columnDelta;
};

static struct Move g_knightMoves[] = {
    {2, 1}, {1, 2}, {-1, 2}, {-2, 1}, {-2, -1}, {-1, -2}, {1, -2}, {2, -1}
};

struct SolutionFinder {
    int rows;
    int columns;
    int * board;
    int maxMoves;
};

#define BOARD_AT(finder, row, column) \
    (*((finder)->board + row * (finder)->columns + column))

static void
initSolutionFinder(struct SolutionFinder * finder, int columns, int rows) {
    int boardSize = rows * columns * sizeof(int);

    finder->columns = columns;
    finder->rows = rows;
    finder->board = malloc((size_t) boardSize);
    memset(finder->board, 0, boardSize);
    finder->maxMoves = rows * columns;
}

static bool
isLegalMove(struct SolutionFinder * finder, int row, int column) {
    return row >= 0 && column >= 0 && row < finder->rows && column < finder->columns &&
        BOARD_AT(finder, row, column) == 0;
}

static bool
makeMove(struct SolutionFinder * finder, int currentRow, int currentColumn, int currentMoveIndex) {
    BOARD_AT(finder, currentRow, currentColumn) = currentMoveIndex;

    if (currentMoveIndex >= finder->maxMoves) {
        // solution found
        return true;
    }

    for (int i = 0; i < 8; ++i) {
        struct Move * move = g_knightMoves + i;
        int nextRow = currentRow + move->rowDelta;
        int nextColumn = currentColumn + move->columnDelta;
        if (isLegalMove(finder, nextRow, nextColumn) &&
            makeMove(finder, nextRow, nextColumn, currentMoveIndex + 1)) {
            return true;
        }
    }

    // this move is not a valid one, reset current mark
    BOARD_AT(finder, currentRow, currentColumn) = 0;
    return false;
}

static void
printBoard(struct SolutionFinder * finder) {
    printf("board:\n");
    for (int row = 0; row < finder->rows; ++row) {
        for (int column = 0; column < finder->columns; ++column) {
            printf(" %3d", BOARD_AT(finder, row, column));
        }
        printf("\n");
    }
}

static void
closeSolutionFinder(struct SolutionFinder * finder) {
    free(finder->board);
}

#define NANO_UNIT       (1000000000LL)

static void printFormattedNanoTime(long long nanoTime, FILE * os) {
    long long sec = nanoTime / NANO_UNIT;
    long long nanos = (nanoTime - (sec * NANO_UNIT));
    fprintf(os, "%lld sec %06lld %03lld msec", sec, nanos / 1000, nanos % 1000);
}

static void
findSolution(int rows, int columns, int startRow, int startColumn) {
    struct SolutionFinder finder;
    struct timeval start;
    struct timeval stop;
    bool found;

    initSolutionFinder(&finder, rows, columns);
    gettimeofday(&start, NULL);
    found = makeMove(&finder, startRow, startColumn, 1);
    gettimeofday(&stop, NULL);

    long long nanoTime = (stop.tv_sec - start.tv_sec) * NANO_UNIT + (stop.tv_usec - start.tv_usec) * 1000L;
    printf("Solution ");
    if (found) {
        printf("found: ");
        printBoard(&finder);
    } else {
        printf("not found");
    }
    printf("\nTime: ");
    printFormattedNanoTime(nanoTime, stdout);
    printf("\n\n");

    closeSolutionFinder(&finder);
}

static int 
getArgParam(char ** argv, int argc, int index, int defaultValue) {
    return index < argc ? atoi(argv[index]) : defaultValue;
}

int main(int argc, char ** argv) {
    int rows = getArgParam(argv, argc, 1, DEFAULT_BOARD_SIDE_SIZE);
    int columns = getArgParam(argv, argc, 2, DEFAULT_BOARD_SIDE_SIZE);
    int iterations = getArgParam(argv, argc, 3, DEFAULT_ITERATIONS);

    printf("KnightTour {rows} {columns} {iterations}\n"
           "Using rows=%d, columns=%d and iterations=%d\n",
           rows, columns, iterations);

    for (int j = 0; j < iterations; ++j) {
        findSolution(rows, columns, DEFAULT_START_ROW, DEFAULT_START_COLUMN);
    }

    return 0;
}

File: src/knight-tour/src/KnightTour.cs
using System;
using System.Text;
using System.Diagnostics;

public sealed class KnightTour
{
    public const int DEFAULT_BOARD_SIDE_SIZE = 8;
    public const int DEFAULT_START_ROW = 0;
    public const int DEFAULT_START_COLUMN = 0;
    public const int MAX_ITERATIONS = 1;

    public static void Main(string[] args)
    {
        var rows = GetArgParam(args, 0, DEFAULT_BOARD_SIDE_SIZE);
        var columns = GetArgParam(args, 1, DEFAULT_BOARD_SIDE_SIZE);

        Console.WriteLine("KnightTour <rows> <columns>\nUsing rows={0}, columns={1}", rows, columns);

        for (var j = 0; j < MAX_ITERATIONS; ++j)
        {
            FindSolution(rows, columns, DEFAULT_START_ROW, DEFAULT_START_COLUMN);
        }
    }

    private static int GetArgParam(string[] args, int index, int defaultValue)
    {
        return index < args.Length ? Convert.ToInt32(args[index]) : defaultValue;
    }

    private static void FindSolution(int rows, int columns, int startRow, int startColumn)
    {
        var finder = new SolutionFinder(rows, columns);
        var watch = new Stopwatch();
        watch.Start();
        var found = finder.Move(startRow, startColumn, 1);
        watch.Stop();
        var elapsed = watch.Elapsed;

        Console.WriteLine("Solution {0}\nTime: {1}",
                          found ? ("Found: " + finder) : "Not Found",
                          elapsed);
    }
}

public sealed class SolutionFinder
{
    private static readonly Move[] KNIGHT_MOVES = new Move[]
    {
        new Move(2, 1), new Move(1, 2), new Move(-1, 2), new Move(-2, 1),
        new Move(-2, -1), new Move(-1, -2), new Move(1, -2), new Move(2, -1)
    };

    private readonly int rows;
    private readonly int columns;
    private readonly int[,] board;
    private readonly int maxMoves;

    public SolutionFinder(int columns, int rows) 
    {
        this.columns = columns;
        this.rows = rows;
        this.board = new int[rows, columns];
        this.maxMoves = columns * rows;
    }

    private bool IsLegalMove(int row, int column)
    {
        return row >= 0 && column >= 0 && row < rows && column < columns && board[row, column] == 0;
    }

    public bool Move(int currentRow, int currentColumn, int currentMoveIndex)
    {
        board[currentRow, currentColumn] = currentMoveIndex;

        if (currentMoveIndex >= maxMoves)
        {
            // solution found
            return true;
        }

        foreach (var move in KNIGHT_MOVES)
        {
            var nextRow = currentRow + move.rowDelta;
            var nextColumn = currentColumn + move.columnDelta;
            if (IsLegalMove(nextRow, nextColumn) && Move(nextRow, nextColumn, currentMoveIndex + 1))
            {
                return true;
            }
        }

        // this move is not a valid one, reset current mark
        board[currentRow, currentColumn] = 0;
        return false;
    }

    public override string ToString()
    {
        var builder = new StringBuilder();
        builder.Append("board:\n");
        for (int row = 0; row < rows; ++row)
        {
            for (int column = 0; column < columns; ++column)
            {
                builder.AppendFormat(" {0, 2}", board[row, column]);
            }
            builder.AppendLine();
        }  
        return builder.ToString();
    }
}

public struct Move
{
    public readonly int rowDelta;
    public readonly int columnDelta;

    public Move(int rowDelta, int columnDelta)
    {
        this.rowDelta = rowDelta;
        this.columnDelta = columnDelta;
    }
}

